psh/pshapp/pshapp.c

```diff
@@ -812,20 +813,95 @@ static int psh_readcmd(struct termios *orig, psh_hist_t *cmdhist, char **cmd)
 }
 
 
-static int psh_parsecmd(char *line, int *argc, char ***argv)
+static char *psh_unescape(char *token, char ch)
 {
-	char *cmd, *arg, **rargv;
+	char *src = token;
+	char *dst = token;
 
-	if ((cmd = strtok(line, "\t ")) == NULL)
+	while (*src != '\0') {
+		if (src[0] == '\\' && src[1] == ch) {
+			*(dst++) = ch;
+			src += 2;
+		}
+		else {
+			*(dst++) = *(src++);
+		}
+	}
+	*dst = '\0';
+	return token;
+}
+
+
+static char *psh_tokenize(char **pstr)
+{
+	char *ptr = *pstr;
+	char *ret = *pstr;
+	char quote = 0;
+
+	if (*ptr == '\0') {
+		return NULL;
+	}
+
+	while (*ptr != '\0' && isspace(*ptr)) {
+		ptr++;
+	}
+
+	ret = ptr;
+
+	if (*ptr == '"' || *ptr == '\'') {
+		quote = *(ptr++);
+		ret = ptr;
+		for (;;) {
+			while (*ptr != '\0' && *ptr != quote) {
+				ptr++;
+			}
+			if (*ptr == '\0' || *(ptr - 1) != '\\') {
+				break;
+			}
+			ptr++;
+		}
+	}
+	else {
+		while (*ptr != '\0' && !isspace(*ptr)) {
+			ptr++;
+		}
+	}
+
+	if (*ptr != '\0') {
+		*(ptr++) = '\0';
+		while (*ptr != '\0' && isspace(*ptr)) {
+			ptr++;
+		}
```

Comment:
I see this was lost somewhere early.
```suggestion
```
You skip the whitespace characters at the beginning of the function.